vncviewer/Win32TouchHandler.cxx

```diff
@@ -0,0 +1,441 @@
+/* Copyright 2020 Samuel Mannehed for Cendio AB
+ *
+ * This is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
+ * USA.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <math.h>
+
+#define XK_MISCELLANY
+#include <rfb/keysymdef.h>
+#include <rfb/Exception.h>
+#include <rfb/LogWriter.h>
+
+#include "i18n.h"
+#include "Win32TouchHandler.h"
+
+static rfb::LogWriter vlog("Win32TouchHandler");
+
+const DWORD MOUSEMOVE_FLAGS = MOUSEEVENTF_MOVE | MOUSEEVENTF_ABSOLUTE |
+                              MOUSEEVENTF_VIRTUALDESK;
+
+const unsigned SINGLE_PAN_THRESHOLD = 50;
+
+Win32TouchHandler::Win32TouchHandler(HWND hWnd) :
+  hWnd(hWnd), gestureStartPos(), fakeButtonMask(0), availabilityLogged(false),
+  gestureActive(false), ignoringGesture(false)
+{
+  // If window is registered as touch we can not receive gestures
+  if (IsTouchWindow(hWnd, NULL)) {
+    throw rfb::Exception(_("Window is registered for touch instead of gestures"));
```

Comment:
If this exception is thrown, it will be caught in touch.cxx and then exit_vncviewer() will be called and an error is logged. This seems to be the correct way forward.